### PR TITLE
Implement vital chef's hat functionality

### PR DESCRIPTION
**Pull request by Tayrtahn**
chefhat

## About the PR

If you put a player-controlled mouse or hamster into a chef's hat and wear it, the mouse/hamster will be able to control your movement. Obviously.

The wearer does still have control as well, so have fun trying to coordinate movement!

## Why / Balance

I noticed that the chef's hat has an inventory that fits a single tiny item, and that mice are tiny items. I tried it and was disappointed to see that this wasn't already a feature. So now it is.

This also enables a potentially friendly interaction between two characters who typically don't get along. As a mouse, do you stop and hope the chef will be kind and put you in his hat instead of the microwave?

## Technical details

Adds PilotedClothingComponent, which is uses InputRelay to set up a connection from the pilot to the wearer. The link is broken if the clothing is unequipped or the pilot leaves the clothing's storage. Both the wearer and the pilot are able to steer the character.

Adds PilotedByClothingComponent to the wearer while the connection is active. This just disables physics prediction for the wearer. Trust me, it feels awful with prediction on. The delay from having it unpredicted adds to the effect anyway.

Adds a tag "ChefPilot" and gives it to mice and hamsters. PilotedClothingSystem uses a whitelist to validate that this tag is on a potential pilot.

## Media

https://github.com/space-wizards/space-station-14/assets/85356/f59d5905-057a-4e17-a220-7ae773c9b040

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- add: Mice and hamsters inside chef's hats can now steer the hat's wearer.

### CODE REVIEW

**Comment by Cojoke-dot** [comment]
This pr... this is a perfect idea...

**Comment by TheShuEd** [comment]
sounds like a rather complex mechanic that is very narrowly applicable and not reusable.  Does your implementation support possible future antagonists like a brain worm or possession as a mechanic for controlling other entities?  Or is it strictly fixed to the conditions of clothing slots

**Comment by ike709** [comment]
> sounds like a rather complex mechanic that is very narrowly applicable and not reusable. 

You have described... Space Station 13

**Comment by Tayrtahn** [comment]
> sounds like a rather complex mechanic that is very narrowly applicable and not reusable.  Does your implementation support possible future antagonists like a brain worm or possession as a mechanic for controlling other entities?  Or is it strictly fixed to the conditions of clothing slots

The way it's implemented, it is limited to the "pilot" being inside a clothing item and the clothing item being worn by the person being controlled.

But the reusable system already existed with the InputRelay system, which is used for transferring control from pilots to mechs. This just bridges a couple of gaps to reuse that functionality for this specific case.

It would definitely be possible to implement a brain worm with InputRelay in a way pretty similar to this (and I might look into that), but I think the differences are significant enough to deserve separate implementations.

**Comment by slarticodefast** [comment]
This idea is amazing. Could you add cockroaches and mothroaches to the pilots as well?

**Comment by beck-thompson** [comment]
Vital PR

**Comment by Tayrtahn** [comment]
> Could you add cockroaches and mothroaches to the pilots as well?

I'm going to keep it limited to rodents for now, but I encourage you or anyone else to submit a follow-up PR for that change if this gets merged.

**Comment by Plykiya** [comment]
mothroaches as hats should let them pilot people as well

**Comment by Hmeister-real** [comment]
> On a side note, if anyone adds a larger chef's hat that can fit a raccoon, they'll have my support instantly

Rat king sized chef hat :godo:

**Comment by asperger-sind** [comment]
Speaking of the prediction, wouldn't this affect players with high ping more than those with less ping? Iirc devbuild has simulated delay of around 185ms so this perhaps didn't come up during testing.

File: Content.Shared/Clothing/EntitySystems/PilotedClothingSystem.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    private void OnEquipped(Entity<PilotedClothingComponent> entity, ref GotEquippedEvent args)
+    {
+        if (!TryComp(entity, out ClothingComponent? clothing))
+            return;
+
+        // Make sure the clothing item was equipped to the right slot, and not just held in a hand.
+        var isCorrectSlot = clothing.Slots.HasFlag(args.SlotFlags);
+        if (!isCorrectSlot)
+            return;
```
Comment (metalgearsloth) [inline Content.Shared/Clothing/EntitySystems/PilotedClothingSystem.cs:63]: HasFlag allocates on debug just manually do a bitmask check.

File: Content.Shared/Clothing/Components/PilotedClothingComponent.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    /// Should click interaction input be relayed from the pilot to the target?
+    /// </summary>
+    /// <remarks>
+    /// This doesn't work very well right now, so it's disabled by default.
+    /// If improvements are made to interaction relays, consider using it.
+    /// </remarks>
+    [DataField]
+    public bool RelayInteraction;
+
+    /// <summary>
```
Comment (metalgearsloth) [inline Content.Shared/Clothing/Components/PilotedClothingComponent.cs:34]: Just remove it if it doesn't work.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I wonder if this would be better served as a generic vehicle system to relay movement inputs instead of having:
- jetpacks
- pilotedclothing
- shuttle
- vehicle 2.0 + cardboard box + mechs.

Specifically by splitting the component into one for being a vehicle and another for being a vehicle able to be piloted via clothing?

File: Content.Shared/Clothing/Components/PilotedClothingComponent.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    /// Reference to the entity contained in the clothing and acting as pilot.
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public NetEntity? Pilot;
+
```
[the remark this answers is not in the document]
Comment (Tayrtahn) [reply Content.Shared/Clothing/Components/PilotedClothingComponent.cs:39]: Huh, didn't know sourcegen could do that. Neat.

**Comment by Tayrtahn** [comment]
> I wonder if this would be better served as a generic vehicle system to relay movement inputs instead of having:
> 
> * jetpacks
> * pilotedclothing
> * shuttle
> * vehicle 2.0 + cardboard box + mechs.
> 
> Specifically by splitting the component into one for being a vehicle and another for being a vehicle able to be piloted via clothing?

You're just trying to dupe me into doing the vehicle refactor, aren't you? 😛 

You make a good point though; all those systems probably could be combined. I can look into doing that, but for now I've just fixed the specific issues you pointed out.

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Comment by GroggleG** [comment]
The completion and merge of this PR is vital to my mental health

**Comment by Kadeo64** [comment]
holy shit SSD people can now be used for something useful

**Comment by metalgearsloth** [comment]
The lack of prediction is pretty bad but I don't think there's an easy way around having 2 players control the same mob so.

**Review by metalgearsloth — APPROVED** [review]

**Comment by justdie12** [comment]
Uploading 2024-07-30 08-12-23.mp4…

**Comment by justdie12** [comment]
Uploading 2024-07-30 08-12-23.mp4…

**Comment by justdie12** [comment]
Uploading 2024-07-30 08-12-23_compressed.mp4…

**Comment by justdie12** [comment]
https://drive.google.com/file/d/1NKXE98TCdvA54PwMnsqiG5EBZmyfXvPm/view?usp=sharing
watch video, i cant upload it here.

**Comment by Cojoke-dot** [comment]
> https://drive.google.com/file/d/1NKXE98TCdvA54PwMnsqiG5EBZmyfXvPm/view?usp=sharing watch video, i cant upload it here.

That is a known bug
#29288

**Comment by justdie12** [comment]
> > https://drive.google.com/file/d/1NKXE98TCdvA54PwMnsqiG5EBZmyfXvPm/view?usp=sharing watch video, i cant upload it here.
> 
> That is a known bug #29288

what about ghost movement? maybe just use a bit of RIPLEY code?

**Comment by Cojoke-dot** [comment]
> what about ghost movement?

I don't think that's a bug and if it was it would not get fixed, its kinda funny and an admin would have to cause it anyway.

**Comment by justdie12** [comment]
> > what about ghost movement?
> 
> I don't think that's a bug and if it was it would not get fixed, its kinda funny and an admin would have to cause it anyway.

i think what author need to use RIPLEY code instead, cuz' now its so laggy tbh
and he needs to add hands and other interface to mouse, in Ratatouille Remi can control hands and oher stuff

**Comment by Cojoke-dot** [comment]
> i think what author need to use RIPLEY code instead, cuz' now its so laggy tbh and he needs to add hands and other interface to mouse, in Ratatouille Remi can control hands and oher stuff

It's using the RIPLEY code, it feels laggy because predictions are turned off so it does not look like your player is teleporting constantly. I agree about that hands but... wyci 🙃

**Pull request merged**